python/tk_multi_workfiles/entity_models/deferred_model.py

```diff
@@ -0,0 +1,588 @@
+                                     so refreshed data is expected later.
+        """
+        if not changed:
+            return
+        parent_item = self.item_from_entity(sg_entity["type"], sg_entity["id"])
+        if not parent_item:
+            return
+        # We use our own deferred cache to avoid conflicts with ShotgunModel so
+        # we can manipulate it freely.
+        parent_uid = parent_item.data(self._SG_ITEM_UNIQUE_ID)
+        if self._deferred_cache.item_exists(parent_uid):
+            # Grab all entries from the iterator
+            existing_uids = [x for x in self._deferred_cache.get_child_uids(parent_uid)]
+        else:
+            existing_uids = []
+
+        sub_entities = []
+        deferred_model = self._deferred_models[sg_entity["id"]]
+        sub_entity_type = deferred_model.get_entity_type()
+        name_field = get_sg_entity_name_field(sub_entity_type)
+
+        # Loop over all entities (leaves) in the deferred model.
+        for sub_entity_id in deferred_model.entity_ids:
+            sub_item = deferred_model.item_from_entity(sub_entity_type, sub_entity_id)
+            if sub_item:
+                sg_data = sub_item.get_sg_data()
+                if sg_data:
+                    sub_entities.append(sg_data)
+        refreshed_uids = []
+        # Ideallly we would retrieve the field list from the ShotgunEntityModel
+        # but this is a private member.
+        deferred_query = self._deferred_query
+        for sub_entity in sub_entities:
+            uids = self._add_deferred_item_hierarchy(
+                parent_item,
+                deferred_query["hierarchy"],
+                name_field,
+                sub_entity,
+            )
+            refreshed_uids.extend(uids)
+        if not sub_entities:
+            # If we don't have any Entity, add a "Retrieving XXXXXs" or
+            # "No XXXXs found" child, depending on `pending_refresh` value.
+            uid = self._add_dummy_place_holder_item(parent_item, pending_refresh)
+            refreshed_uids.append(uid)
+        # Go through the existing items and discard the ones which shouldn't be
+        # there anymore.
+        for uid in existing_uids:
+            if uid in refreshed_uids:
+                # Here we could update items from the refreshed data
+                continue
+            data_item = self._deferred_cache.take_item(uid)
+            item = self._get_item_by_unique_id(uid)
+            if item:
+                # The item might be already gone because one its parent was
+                # deleted.
+                self._delete_item(item)
+
+    @staticmethod
+    def _deferred_entity_uid(sg_entity):
+        """
+        Returns a unique id for the given Entity retrieved in a deferred query.
+        """
+        # ShotgunModel uses the entity id for leaves, we use the Entity type and
+        # its id to avoid clashes in the various internal caches.
```

Comment:
are we sharing a space with internal caches? Is that bad encapsulation?

Reply:
We don't really share any internal cache, but when adding items in the primary model uids are added in `ShotgunQueryModel.___items_by_uid` for fast lookup. Should I update the comment?